Fix #19578: Allow `spack clean` to remove cached resources

teaguesterling opened the pull request:
This is a naive fix to address #19578 by adding an additional `spack.stage.purge_resources` function and corresponding args to the `spack.cmd.clean` CLI. It may be better to unify the `purge` and `purge_resources` functions into one, but this was the least invasive way to address the issue without changing function signatures.

An alternative and less repetitive implementation of `spack.stage.purge` that would not change default behaviors would be https://github.com/teaguesterling/spack/blob/ISSUE-19578-alt-impl/lib/spack/spack/stage.py#L953. I'm not yet familiar enough with Spack coding standards sure which is more in-line with convention.  
```
def purge(builds=True, resources=False):
    """Remove all build and/or resource directiories in the top-level stage path."""
    root = get_stage_root()
    if os.path.isdir(root):
        for stage_dir in os.listdir(root):
            is_build_stage = stage_dir.startswith(stage_prefix) or stage_dir == ".lock"
            is_resource_stage = stage_dir.startswith(resource_prefix)
            if builds and is_build_stage or resources and is_resource_stage:
                stage_path = os.path.join(root, stage_dir)
                if os.path.isdir(stage_path):
                    remove_linked_tree(stage_path)
                else:
                    os.remove(stage_path)
```

teaguesterling committed:
Adding initial resource cleanup

teaguesterling committed:
Adding resource purging

haampie commented:
I wouldn't distinguish between stage dirs and resource dir. Staging sets up both, so `clean -s` should clean up both.

Can't `stage_prefix` be used?

teaguesterling commented:
> I wouldn't distinguish between stage dirs and resource dir. Staging sets up both, so `clean -s` should clean up both.
> 
I like this as well, but wasn't sure of what would be preferred.

> Can't `stage_prefix` be used?

The handling of resources staging is a bit funky as far as I could follow. This is my first time looking in the Spack source. The "resource-" bit being hard coded when created. Wouldn't be a bad idea to put that in a constant as well.

teaguesterling committed:
Update stage.py

Simplify cleaning process

teaguesterling committed:
Update clean.py

Remove separate clean resource command options

teaguesterling committed:
Making sure resource_prefix stays consistent between stage and package_base
It may eventually make sense to refactor `_resource_stage` and the
corresponding logic out of `package_base` and into `stage`.

teaguesterling committed:
Update clean.py

haampie commented:
@spackbot fix style

teaguesterling committed:
[@spackbot] updating style on behalf of teaguesterling

teaguesterling committed:
Merge branch 'develop' into ISSUE-19578

teaguesterling committed:
Update stage.py

teaguesterling committed:
Update stage.py to simplify logic
